// @sme.up/kokos-sdk-node/lib/services/JS_00_02.js
const JS_00_01 = {
    methods: {
        "GET.TRE": getTree,
        "GET.EXB": getTable,
        "THR.ERR": throwError,
    },
};
async function getTree(_fun, _context, printer) {
    printer.writeDataNode({
        children: [],
        obj: {
            t: "CN",
            p: "COL",
            k: "PIPPO",
        },
        value: "Pippo",
    });
}
async function getTable(_fun, _context, printer) {
    printer.writeDataColumns([
        {
            name: "COL1",
            title: "Column 1",
        },
        {
            name: "COL2",
            title: "Column 2",
        },
    ]);
    printer.writeDataRow({
        cells: {
            COL1: {
                value: "Pippo",
                obj: {
                    t: "CN",
                    p: "COL",
                    k: "PIPPO",
                },
            },
            COL2: {
                value: "Pluto",
                obj: {
                    t: "CN",
                    p: "COL",
                    k: "PLUTO",
                },
            },
        },
    });
}
async function throwError(_fun, _context, _printer) {
    throw new Error("Dummy error");
}
export default JS_00_01;
