// @sme.up/kokos-sdk-node/lib/types/data-structures/smeupDataTable.js
/**
 * Represent the possible shapes of a field
 */
export var SmeupCellShapes;
(function (SmeupCellShapes) {
    SmeupCellShapes["AUTOCOMPLETE"] = "ACP";
    SmeupCellShapes["BUTTON_LIST"] = "BTN";
    SmeupCellShapes["BUTTON"] = "BTN";
    SmeupCellShapes["CHART"] = "GRA";
    SmeupCellShapes["CHECKBOX"] = "CHK";
    SmeupCellShapes["CHIP"] = "CHI";
    SmeupCellShapes["COLOR_PICKER"] = "CLP";
    SmeupCellShapes["COMBOBOX"] = "CMB";
    SmeupCellShapes["EDITOR"] = "EDT";
    SmeupCellShapes["GAUGE"] = "GAU";
    SmeupCellShapes["IMAGE"] = "IMG";
    SmeupCellShapes["INPUT_CHECKBOX"] = "INC";
    SmeupCellShapes["INPUT_FIELD"] = "INF";
    SmeupCellShapes["KNOB"] = "KNB";
    SmeupCellShapes["MULTI_AUTOCOMPLETE"] = "AML";
    SmeupCellShapes["MULTI_COMBOBOX"] = "CML";
    SmeupCellShapes["PROGRESS_BAR"] = "PGB";
    SmeupCellShapes["RADIO"] = "RAD";
    SmeupCellShapes["RATING"] = "RTG";
    SmeupCellShapes["SWITCH"] = "SWT";
    SmeupCellShapes["TEXT_FIELD"] = "ITX";
})(SmeupCellShapes || (SmeupCellShapes = {}));
